climada_petals/engine/warn.py

```diff
@@ -0,0 +1,191 @@
+                if self.filter_data.operations[i] == 'DILATION':
+                    reg_curr_lvl = skimage.morphology.dilation(reg_curr_lvl,
+                                                               skimage.morphology.disk(self.filter_data.sizes[i]))
+                elif self.filter_data.operations[i] == 'EROSION':
+                    reg_curr_lvl = skimage.morphology.erosion(reg_curr_lvl,
+                                                              skimage.morphology.disk(self.filter_data.sizes[i]))
+                elif self.filter_data.operations[i] == 'MEDIANFILTERING':
+                    filter_med = np.ones((self.filter_data.sizes[i], self.filter_data.sizes[i]))
+                    reg_curr_lvl = skimage.filters.median(reg_curr_lvl, filter_med)
+                else:
+                    LOGGER.warning("The operation is not defined. "
+                                   "Please select 'EROSION', 'DILATION', or 'MEDIANFILTERING'.")
+            return reg_curr_lvl
+
+        max_warn_level = np.max(d_thrs)
+        if max_warn_level == 0:
+            return np.zeros(d_thrs.shape)
+
+        warn_regions = 0
+        # iterate from highest level to lowest (0 not necessary, because rest is level 0)
+        for j in range(max_warn_level, 0, -1):
+            w_l = filtering(d_thrs, warn_regions, j)
+            # keep warn regions of higher levels by taking maximum
+            warn_regions = np.maximum(warn_regions, w_l)
+        self.warning = warn_regions
+
+    @staticmethod
+    def remove_small_regions(warning, size_thr):
+        """Remove regions smaller than defined threshold from warning.
+
+        Parameters
+        ----------
+        warning : np.array
+            Warning where regions should be removed from.
+        size_thr : int
+            Threshold defining too small regions (number of grid points).
+
+        Returns
+        ----------
+        warning : np.array
+            Warning without too small regions.
+        """
+
+        def increase_levels(warn, size):
+            # increase levels of too small regions to max level of this warning
+            labels = skimage.measure.label(warn)
+            for i in range(np.max(labels)):
+                cnt = np.count_nonzero(labels == i)
+                if cnt <= size:
+                    warn[labels == i] = np.max(warn, axis=(0, 1))
+            return warn
+
+        def set_new_lvl(warn, size):
+            # correct the max_lvl regions generated before down,
+            # until the new regions with it are large enough
+            for i in range(np.max(warn, axis=(0, 1)), np.min(warn, axis=(0, 1)), -1):
+                level = copy.deepcopy(warn)
+                level[warn != i] = 0
+                labels = skimage.measure.label(warn)
+                for j in range(len(np.unique(labels)) + 1):
+                    cnt = np.count_nonzero(labels == j)
+                    if cnt <= size:
+                        warn[labels == j] = i - 1
```

Comment:
This method return nothing... modifying input variables in-place is not a very good idea.